src/core/vanilla.ts

```diff
@@ -396,129 +396,80 @@ const invalidateDependents = <Value>(state: State, atom: Atom<Value>): void => {
 const writeAtomState = <Value, Update>(
   state: State,
   atom: WritableAtom<Value, Update>,
-  update: Update,
-  pendingPromises: Promise<void>[]
+  update: Update
 ): void => {
-  const isPendingPromisesExpired = !pendingPromises.length
   const writePromise = getAtomState(state, atom)?.w
   if (writePromise) {
-    const promise = writePromise.then(() => {
-      writeAtomState(state, atom, update, pendingPromises)
-      if (isPendingPromisesExpired) {
-        flushPending(state)
-      }
+    writePromise.then(() => {
+      writeAtomState(state, atom, update)
+      flushPending(state)
```

Comment:
Don't we need to guard for expired promises anymore?

Reply:
No, because we don't return a promise. There's no concept of expiration.
So, this was really confusing in code. Nobody understands what's going on. Hence I'm willing to make this change before v1, and if we have real requirement, let's consider a possible solution with better code.